### PR TITLE
Explicitly state minimum NodeJS version supported

### LINKED ISSUE #25: Minimum supported version of Node.js
What is the minimum supported version of Node.js?

### LINKED ISSUE #30: package-lock.json is out of date
615f181 bumped the version of the package in package.json but did not update package-lock.json accordingly.

### PR DESCRIPTION
**Pull request by robertodauria**
This PR makes it explicit which version of NodeJS is required to run the ndt7 client. NodeJS < 10 does not support URL and probably other features that the code uses.

This also allows us to remove the dependency on workerjs, as Node >= 10.5.0 has native support for workers via the `workers_thread` module.

Closes #25 
Closes #30 

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/ndt7-js/46)

### CODE REVIEW

**Comment by RA80533** [comment]
Should we look at raising the minimum version to Node.js 12? According to the official Node.js release schedule, Node.js 10 has been EOL since the release of Node.js 16 on 2021-04-20, making Node.js 12 the oldest LTS version.

Users looking to use an EOL version can always install the package with `--ignore-engines` or an equivalent flag based on their package manager of choice and patch whatever APIs they need.

**Comment by robertodauria** [comment]
I had just looked at which features we use / what's the minimum version of NodeJS that supports them. If the users can still use EOL versions if they really need to then defaulting to >=12 sounds good! Thanks.

File: package-lock.json
Comment (RA80533) [inline package-lock.json:4252]: Did simply regenerating package-lock.json cause this property to be created? If so, sshpk may not have been installed correctly when it was originally added to the project.

**Comment by RA80533** [comment]
.travis.yml should be updated accordingly to reflect whatever version is eventually decided upon:

https://github.com/m-lab/ndt7-js/blob/fa6a78dff2c1ff391581db62cc957ce113fed406/.travis.yml#L3-L4

File: package-lock.json
Comment (robertodauria) [reply package-lock.json:4252]: Yes, I didn't change anything manually in package-lock.json. Perhaps it wasn't being created with npm 6.x and it's an effect of switching to npm 7.x with `nvm use 12`?

Comment (RA80533) [reply package-lock.json:4252]: Hm, it's possible.

What version of npm did you use? I can't seem to actually replicate it.

File: package.json
Comment (RA80533) [inline package.json:27]: - If we want to support Node.js 12 in general, it might just be better to declare `>=12`.
- If 12.13.0 was chosen because it was the first LTS version of Node.js 12, the ideas conveyed in the documentation should be shifted to reflect the slight difference therein ("Node.js 12" to "LTS versions of Node.js 12" or something of that nature).

File: package-lock.json
Comment (robertodauria) [reply package-lock.json:4252]: Last time package-lock.json was updated was in PR #35, but just running `npm install` on the current `main` branch after removing package-lock.json re-creates this section. It's totally possible that sshpk (which I'm not familiar with) wasn't installed correctly to begin with, but I don't know why previous runs of `npm install` didn't add this section before.

Maybe it's been brought in by some minor version update on some dependency? The dependency chain is:

```
- ├─┬ coveralls@3.1.0
│ └─┬ request@2.88.2
│   ├─┬ http-signature@1.2.0
│   │ └─┬ sshpk@1.16.1
```

coveralls has been at version 3.1.0 for a year now, request@2.88.2 is pretty old as well and http-signature@1.2.0 is 4 years old.

Comment (RA80533) [reply package-lock.json:4252]: I forgot that npm attempts to reconcile dependency versions with respect to those currently residing in the local node_modules folder. Once I removed that folder and attempted to install everything again, I had a new package-lock.json with that property.

Comment (robertodauria) [reply package-lock.json:4479]: I think I found why.

If I just run
```
git checkout main
rm package-lock.json
npm install
```

That bin section is re-created, regardless of which npm version I am using. But if I run:

```
nvm use 10
rm package-lock.json
rm -rf node_modules
npm install
```

There isn't any bin section anymore. I think my node_modules might have been created with npm v7 (that comes with node v12), which leads to the creation of that section even if using an older npm to run `npm install` later.

I've pushed an update of a new, clean package-lock.json after removing `node_modules` and just running `npm install` once with npm v7.

File: package.json
Comment (robertodauria) [reply package.json:27]: Do you have any recommendations on which one would be better? I'm personally fine with either, as long as it's a version supporting all the features we use in the client. I'm not very familiar with npm packaging in general, so I'm happy to defer to your judgment here.

**Comment by RA80533** [comment]
```diff
diff --git a/src/ndt7.min.js b/src/ndt7.min.js
index 957ddd1..2b0e47c 100644
--- a/src/ndt7.min.js
+++ b/src/ndt7.min.js
@@ -1 +1 @@
-[…]
+Promise { <pending> }
```

File: package-lock.json
Comment (RA80533) [reply package-lock.json:4497]: > There isn't any bin section anymore. I think my node_modules might have been created with npm v7 (that comes with node v12), which leads to the creation of that section even if using an older npm to run npm install later.

Yeah, npm seems to trust and reconcile what it already in node_modules with whatever is in package-lock.json, even if package-lock.json is outdated (like in this case). I did the exact thing as you.

Keep in mind that some Node.js versions allow for newer versions of npm. The latest version of Node.js 12 (`v12.22.1`) comes bundled with npm 6.x but that npm can be updated to the latest major version—7.x (which resolves to 7.13.0). Since `nvm use <version>` may or may not correspond to a specific version of npm, it's important to isolate its version as much as possible. In most cases, running `npx npm@<version> […]` will be enough to test various behaviors for any particular version of npm as long as the npx utility you're working with supports running npm in that way.

> I've pushed an update of a new, clean package-lock.json[…]

The new package-lock.json will definitely alleviate the reproducibility issues we're seeing here.

**Comment by robertodauria** [comment]
> ```diff
> diff --git a/src/ndt7.min.js b/src/ndt7.min.js
> index 957ddd1..2b0e47c 100644
> --- a/src/ndt7.min.js
> +++ b/src/ndt7.min.js
> @@ -1 +1 @@
> -[…]
> +Promise { <pending> }
> ```
> 
> It appears that an asynchronous operation was used to produce the contents of the output file but it was never awaited upon.

Ah, interesting. It seems that comes from minify (`minify --js < src/ndt7.js`), rather than our code. Updating to a more recent version of minify (7.0.2) fixed it, going back to the previous version breaks it again. The most recent versions of minify require node >= 14, but minify 6.x works with node 12 and does not cause this issue. I've updated it, which has fixed the build.

**Comment by RA80533** [comment]
Based on the changes made to package-lock.json, this PR supersedes #31 and should therefore close #30 once merged. Additionally, the superseded PR will have to be manually closed.

It also appears that the PR will resolve the issue for which #5 was created to fix.

**Comment by robertodauria** [comment]
Agreed. I've changed the PR description to close #30 and will close #31 and #5 manually once it's merged.

Re. the value for `engines > node`, I've set it to `>=12` as that seems to be what we want, compatibility-wise.

Any further thoughts on this PR?

**Comment by RA80533** [comment]
I've never seen an engine version declared specifically to its initial LTS version but I find it to be quite informative in its intention.

Sourcegraph shows 34+ results for `>=12.13.0` ([query](https://sourcegraph.com/search?q=lang:json+%22node%22:+%22%3E%3D12.13.0%22&patternType=literal)) compared to 500+ results for `>=12` ([query](https://sourcegraph.com/search?q=lang:json+%22node%22:+%22%3E%3D12%22&patternType=literal)). While `>=12` is undoubtably the conventional way projects have declared their support for Node.js 12 (LTS or not), an exact minimum version is beneficial in explicitly providing contributors with a version to develop and test against.

If it isn't too close to the bike shed in our backyard, I believe there is merit in declaring `>=12.13.0` as long as our tooling supports it. Off the top of my head, we may need to make an accommodation for Travis CI because its configuration file only takes into account major versions for `node_js`.

File: package.json
Comment (robertodauria) [reply package.json]: If I understand your [subsequent comment](https://github.com/m-lab/ndt7-js/pull/46#issuecomment-842773115) correctly, `>= 12` and the current phrasing in the README seems ok.

**Comment by robertodauria** [comment]
@RA80533 got it, I left it as `>= 12`.

As a general policy, we never merge a PR without explicit approval. Would you be willing to approve this one?

**Review by RA80533 — APPROVED** [review]

**Comment by RA80533** [comment]
GitHub’s UI isn’t very clear on this but I believe required reviewers must have write permissions to the repository for their approvals to count. I did, however, approve the changes nonetheless.

**Pull request merged**
